modules/transport-netty4/src/test/java/org/elasticsearch/transport/netty4/Netty4WriteThrottlingHandlerTests.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+package org.elasticsearch.transport.netty4;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.embedded.EmbeddedChannel;
+
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.test.ESTestCase;
+import org.junit.After;
+import org.junit.Before;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutionException;
+
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.lessThan;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+
+public class Netty4WriteThrottlingHandlerTests extends ESTestCase {
+
+    private SharedGroupFactory.SharedGroup transportGroup;
+
+    @Before
+    public void createGroup() {
+        final SharedGroupFactory sharedGroupFactory = new SharedGroupFactory(Settings.EMPTY);
+        transportGroup = sharedGroupFactory.getTransportGroup();
+    }
+
+    @After
+    public void stopGroup() {
+        transportGroup.shutdown();
+    }
+
+    public void testThrottlesLargeMessage() throws ExecutionException, InterruptedException {
+        final List<ByteBuf> seen = new CopyOnWriteArrayList<>();
+        final EmbeddedChannel embeddedChannel = new EmbeddedChannel(
+            new CapturingHandler(seen),
+            new Netty4WriteThrottlingHandler(new ThreadContext(Settings.EMPTY))
+        );
+        final int writeableBytes = Math.toIntExact(embeddedChannel.bytesBeforeUnwritable());
+        assertThat(writeableBytes, lessThan(Netty4WriteThrottlingHandler.MAX_BYTES_PER_WRITE));
+        final int fullSizeChunks = randomIntBetween(2, 10);
+        final int extraChunkSize = randomIntBetween(0, 10);
+        final ByteBuf message = Unpooled.wrappedBuffer(
+            randomByteArrayOfLength(Netty4WriteThrottlingHandler.MAX_BYTES_PER_WRITE * fullSizeChunks + extraChunkSize)
+        );
+        final ChannelPromise promise = embeddedChannel.newPromise();
+        transportGroup.getLowLevelGroup().submit(() -> embeddedChannel.write(message, promise)).get();
+        assertThat(seen, hasSize(1));
+        assertEquals(message.slice(0, Netty4WriteThrottlingHandler.MAX_BYTES_PER_WRITE), seen.get(0));
+        assertFalse(promise.isDone());
+        transportGroup.getLowLevelGroup().submit(embeddedChannel::flush).get();
```

Comment:
Flush always flushes fully since we use the embedded channel. Can we add a test that verifies that throttling in`doFlush` works, perhaps like `Netty4HttpServerTransportTests.testLargeResponse` (but without the pipelining handler)?

Reply:
> Flush always flushes fully since we use the embedded channel. 

It turns out that the embedded one doesn't really behave differently and will trigger a series of flushes like a real channel :)
=> I did the same test (in principle) I did for HTTP here now as well for the large message sending case and asserted that we do in fact see the throttling.